src/Features/Core/Portable/Diagnostics/DiagnosticHelper.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal static class DiagnosticHelper
+    {
+        /// <summary>
+        /// Creates a <see cref="Diagnostic"/> instance.
+        /// </summary>
+        /// <param name="descriptor">A <see cref="DiagnosticDescriptor"/> describing the diagnostic.</param>
+        /// <param name="location">An optional primary location of the diagnostic. If null, <see cref="Location"/> will return <see cref="Location.None"/>.</param>
+        /// <param name="effectiveSeverity">Effective severity of the diagnostic.</param>
+        /// <param name="additionalLocations">
+        /// An optional set of additional locations related to the diagnostic.
+        /// Typically, these are locations of other items referenced in the message.
+        /// If null, <see cref="Diagnostic.AdditionalLocations"/> will return an empty list.
+        /// </param>
+        /// <param name="properties">
+        /// An optional set of name-value pairs by means of which the analyzer that creates the diagnostic
+        /// can convey more detailed information to the fixer. If null, <see cref="Diagnostic.Properties"/> will return
+        /// <see cref="ImmutableDictionary{TKey, TValue}.Empty"/>.
+        /// </param>
+        /// <param name="messageArgs">Arguments to the message of the diagnostic.</param>
+        /// <returns>The <see cref="Diagnostic"/> instance.</returns>
+        public static Diagnostic Create(
+            DiagnosticDescriptor descriptor,
+            Location location,
+            ReportDiagnostic effectiveSeverity,
```

Comment:
This concept just doesn't make sense to me. By design a Diagnostic doesn't have an effective severity and there is no way to know what the effective severity will be until all options are taken into account. It feels like this API is a path of failure.

Reply:
> By design a Diagnostic doesn't have an effective severity and there is no way to know what the effective severity will be until all options are taken into account.

This is not correct in practice. For example, the naming style rules can be configured so naming style violations are reported at different severities in different locations, even within the same file. This is certainly at odds with the way rule set files are structured, and attempting to set the severity of a diagnostic within the rule set file is a known way unintentionally disable .editorconfig's finer-grained severity settings.

@mavasani has requested I create a follow-up issue with a proposal that transitions from per-diagnostic effective severity to "sub-diagnostics" (https://github.com/dotnet/roslyn/pull/27220#discussion_r193239176), but until that change is implemented diagnostics need an effective severity for feature parity.

Reply:
I think we're using similar terms with different meanings. What you call "effectiveSeverity" is not necessarily what the severity will be when `GetEffectiveDiagnostics` is called. I would instead just call the parameter "severity" or something similar. You're basically creating multiple diagnostics with the same ID and different default severity.